File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> window; // To store unique characters in the current window
        int i = 0, j = 0; // Sliding window pointers
        int maxi = 0; // Maximum length of substring without repeating characters

        while (j < s.size()) {
            // If the character is not in the window, expand the window
            if (window.find(s[j]) == window.end()) {
                window.insert(s[j]);
                maxi = max(maxi, j - i + 1); // Update the maximum length
                j++;
            } 
            // If the character is already in the window, shrink the window from the left
            else {
                window.erase(s[i]);
                i++;
            }
        }

        return maxi;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        Set=set()
        left=right=0
        if len(s)==0:
            return 0
        length=-1
        while right<len(s)-1:
            while Set and s[right] in Set:
                Set.remove(s[left])
                left+=1
            Set.add(s[right])
            length=max(length,right-left+1)
            right+=1
        return length
        